crates/solidity/outputs/cargo/crate/src/backend/passes/p5_resolve_references.rs

```diff
@@ -1328,15 +2235,85 @@ impl Visitor for Pass {
+                    })
+                    .and_then(|reference| reference.resolution.as_definition_id());
+                self.resolve_named_arguments(&named_arguments.arguments, definition_id);
+            }
+        }
+    }
+
+    fn leave_variable_declaration_statement(
+        &mut self,
+        node: &input_ir::VariableDeclarationStatement,
+    ) {
+        if self.language_version >= VERSION_0_5_0 {
+            // in Solidity >= 0.5.0 we need to update the scope so further
+            // resolutions can access this variable definition
+            self.enter_replace_scope_for_node_id(node.node_id);
+            // NOTE: ensure following code does not need to perform resolution
+        }
+
+        if matches!(
+            node.variable_type,
+            input_ir::VariableDeclarationType::VarKeyword
+        ) {
+            // update the type of the variable with the type of the expression (if available)
+            if let Some(value) = &node.value {
+                let typing = self
+                    .typing_of_expression(&value.expression)
+                    .as_type_id()
+                    .map_or(Typing::Unresolved, |type_id| {
+                        Typing::Resolved(self.types.reified_type(type_id))
+                    });
+                self.binder.fixup_node_typing(node.node_id, typing);
+            }
+        }
+    }
+
+    fn leave_tuple_deconstruction_statement(
+        &mut self,
+        node: &input_ir::TupleDeconstructionStatement,
+    ) {
+        if self.language_version >= VERSION_0_5_0 {
+            // in Solidity >= 0.5.0 we need to update the scope so further
+            // resolutions can access these variable definitions
+            self.enter_replace_scope_for_node_id(node.node_id);
+            // NOTE: ensure following code does not need to perform resolution
+        }
+
+        if node.var_keyword.is_none() {
+            return;
+        }
+
+        let typing = self.typing_of_expression(&node.expression);
+        let Typing::Resolved(tuple_type_id) = typing else {
+            // we can't fixup typing if the expression failed to type
+            return;
+        };
+        let types = if let Type::Tuple { types } = self.types.get_type_by_id(tuple_type_id) {
+            types.clone()
+        } else {
+            // the resolved type is not a tuple
+            vec![tuple_type_id]
+        };
+
+        // fixup typing of `var` declarations
+        for (element, element_type_id) in node.elements.iter().zip(types) {
+            let Some(member) = &element.member else {
+                continue;
+            };
+            if let input_ir::TupleMember::UntypedTupleMember(untyped_tuple_member) = member {
+                let typing = Typing::Resolved(self.types.reified_type(element_type_id));
+                self.binder
+                    .fixup_node_typing(untyped_tuple_member.node_id, typing);
+            }
         }
```

Comment:
not blocking: as this pass is almost doubling in size due to argument lookup/overload resolution, I wonder if the module should be split into submodules for readability/isolation?
Is the logic is independent of other steps in the same pass? I can only see it growing in the future as we add analysis of literal types/strings, and other `TODO` parts. The main pass can still call into separate "sub-resolvers" as needed.

Reply:
I refactored and split this pass `p5_resolve_references` as well as `p4_type_definitions` and the `types` module as you suggested. This results in an increased PR size because of all the moved code. Happy to extract the relevant commits for a later PR if it makes reviewing easier.